lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,277 @@
+
+    PLUGIN_PROPERTIES = "plugin-descriptor.properties"
+    CONFIG_YML = "opensearch.yml"
+
+    def __init__(self, name: str, charm: Object, relname: Optional[str] = None):
+        """Creates the OpenSearchPlugin object.
+
+        The *args enable children classes to pass relations.
+        """
+        self.relname = relname
+        self.charm = charm
+        self.distro = self.charm.opensearch
+        self.CONFIG_YML = self.charm.opensearch_config.CONFIG_YML
+        self._name = name
+        self._plugin_properties = Properties()
+
+    @property
+    def version(self) -> str:
+        """Returns the current version of the plugin."""
+        with open(os.path.join(f"{self.distro.paths.plugins}", f"{self.PLUGIN_PROPERTIES}")) as f:
+            self._plugin_properties.load(f.read())
+        return self._plugin_properties._properties["version"]
+
+    def _is_started(self) -> bool:
+        return self.distro.is_started()
+
+    def _request(self, *args, **kwargs) -> dict:
+        return self.distro.request(*args, **kwargs)
+
+    def is_related(self) -> bool:
+        """Returns True if a relation is expected and it is set."""
+        if not self.relname:
+            return True
+        return len(self.charm.framework.model.relations[self.relname] or {}) > 0
+
+    def _update_keystore_and_reload(
+        self, keystore: Dict[str, str], force: bool = True, remove_keys: bool = False
+    ) -> None:
+        if not keystore:
+            return
+        try:
+            for key, value in keystore.items():
+                if remove_keys:
+                    self.distro.remove_from_keystore(key)
+                else:
+                    self.distro.add_to_keystore(key, value, force=force)
+            # Now, reload the security settings and return if opensearch needs restart
+            post = self._request("POST", "_nodes/reload_secure_settings")
+            logger.debug(f"_update_keystore_and_reload: response received {post}")
+        except OpenSearchKeystoreError as ek:
+            raise ek
+        except Exception as e:
+            logger.exception(e)
+            raise OpenSearchPluginError("Unknown error during keystore reload")
+        if post["status"] < 200 or post["status"] >= 300:
+            raise OpenSearchPluginError("Error while processing _nodes reload")
+
+    @property
+    def name(self) -> str:
+        """Returns the name of the plugin."""
+        return self._name
+
+    def uninstall(
+        self,
+        opensearch_yml: Dict[str, str],
+        keystore: Dict[str, str],
+    ) -> bool:
+        """Erases relevant data for this plugin. Returns True if restart needed."""
+        if not self._is_started():
+            return False
+        self._update_keystore_and_reload(keystore)
+        return any(
+            [
+                self.distro.config.delete(self.CONFIG_YML, c, v)[c] == v
+                for c, v in opensearch_yml.items()
+            ]
+        )
```

Comment:
This is the job of the plugin manager (while solely interacting with a method in `OpenSearchConfig` and not the yaml filte itself)

Reply:
We need to do the following changes:
* Move away from self.distro to self.charm.opensearch_config
* Create a proxy method to put and delete a collection of entries in OpenSearchConfig. That method should receive a dict containing the key/values.